Mcrf profile url replace

rbaugh opened the pull request:
Using the `{rest_api_url}` merge tag in the webhook url, will now get replaced with the MCRF profile url when available. I had to switch to the `gform_pre_replace_merge_tags` hook as the WebHooks addon was using this and needed to be able to replace first. Fallback is to leave as is and let the WebHooks replacement pick up if CiviCRM is installed locally.

There was some type casting warnings I saw and tried to add checks for. Some extra formatting changes due to my dev environment.

rbaugh commented:
Initial code was just to replace the `{rest_api_url}` merge tag, but realized that there could be other forms that may want to post to the WP rest API. So I added in a settings for the form with an option to use the MCRF connection in tandem with the form. If set, it will use either the default profile found in the main settings or the one chosen on the form.

rbaugh commented:
FYI, Form settings `dependency` property isn't like normal conditionals. So once you select the checkbox to "Use the MCRF", you must save those changes before prompted to select the profile. Likely could be enhanced with some JS, but haven't dug into that just yet.

rbaugh commented:
Wondering if this should get moved to the WebHook feed settings. So you can configure per WebHook feed entry. Thoughts?

agileware-justin commented:
> Wondering if this should get moved to the WebHook feed settings. So you can configure per WebHook feed entry. Thoughts?

This is an important point. Because replacing {rest_api_url} would impact all Webhook feeds using that merge tag. Which may not be correct in all instances - someone out there may be using Webhook feeds to send data to WordPress as well as CiviCRM. It could happen and there would be complaints.

With that in mind, it would probably be a good idea to instead create a new merge tag, something like ```{civicrm_rest_api_url}```

That would avoid any potential future issues and complaints.

This ```{civicrm_rest_api_url}``` could then be populated with either:

* The same value as: ```{rest_api_url}civicrm/v3/rest``` - If no CiviCRM McRF
* CiviCRM McRF rest end point - If using CiviCRM McRF

It's also been a bit fiddly to have to enter ```{rest_api_url}civicrm/v3/rest``` each time in the webook URL and potentially problematic. This would make things a bit simpler.

Shortening a bit the current long URLs like:
```{rest_api_url}civicrm/v3/rest?entity=FormProcessor&action=newsletter_subscribe&key={gf_civicrm_site_key}&api_key={gf_civicrm_api_key}&json=1```

To:
```{civicrm_rest_api_url}?entity=FormProcessor&action=newsletter_subscribe&key={gf_civicrm_site_key}&api_key={gf_civicrm_api_key}&json=1```

Back to the WebHook feed settings, I think it's something to keep in mind that Gravity Forms on a site may use different WebHook feed endpoints for different forms, we have customers that definitely do use it this way. Sending some Gravity Forms submissions to MS Dynamics and others to CiviCRM. There is also the flexibility built into Gravity Forms to allow a single Gravity Form to have multiple WebHook feeds. Both these use cases need to be kept in mind.

To that end, we've (Agileware team) discussed replacing the entire WebHook feed plugin with a custom CiviCRM feed plugin to make the integration easier to set up. That's only an idea at this stage for some future development. Just sharing in case you were interested.

agileware-justin commented:
Thanks for having a go and submitting the PR too! Exciting to see that someone else is trying out this plugin.

rbaugh commented:
Currently this PR will add another tab on the Form settings screen. Here you can choose to use the CiviCRM setting to override the `{rest_api_url}` in the webhook currently for the form itself. But that had me thinking that maybe hooking into the WebHook feed settings to append this checkbox and profile selection might be better so that the override is per webhook and not on the form itself.

This would allow for having more than one WebHook per form and it would then allow more finite control over the WebHook in case you wanted to send the form data to multiple CiviCRM instances.

I can look at moving it to the WebHook Feed settings/fields. This way you don’t have to worry about replicating the same features as the WebHook in order to have your own Feed in this plugin. I mean it doesn’t hurt to have all functionality, outside of GF itself, in the one plugin. Shouldn’t take too long to replicate the inner works of the WebHooks plugin into here.

agileware-iris commented:
Had a look - looks alright so far! Some changes required though.

To summarise the current state, I think the following are still todo/need modifying.

- [ ] `{rest_api_url}` should work as before, replaced with the current domain's REST API url (see https://docs.gravityforms.com/rest_api_url-merge-tag/).
- [ ]  Add a new merge tag `{civicrm_rest_api_url}` for when we explicitly want to define REST endpoints via CMRF. I believe this is just a matter of changing the tag in `replace_merge_tags()`.
- [ ] Change all references from MCRF (McRestFace) to CMRF (CiviMcRestFace) for consistency.
- [ ]  Move the form setting "Use CiviCRM REST Connection Profile" to webhook feeds settings, so it can be enabled on a per-feed basis.

Would you be able to make those changes, @rbaugh?

It might be a good idea to move the fourth item (moving to webhook settings) to a new PR, to keep the scope of this one small.

rbaugh commented:
@agileware-iris 
I just pushed up some changes that move the MCRF settings into the WebHooks settings. Since this plugin currently relies on the WebHooks AddOn, I wanted to be able to set a profile per feed entry.

I also added a new merge tag `{civicrm_mcrf_url}` for use to not conflict with the WP API url. The tag will be replaced with the url from the MCRF profile and also the site key and api key params. No need to add those into the WebHook feed since we know what the structure should be and will set them based on the feed set in WebbHook or defaults back to plugin setting. The url would now be formatted as `{civicrm_mcrf_url}&entity=FormProcessor&action=FORM_NAMEjson=1` and you just need to replace the `FORM_NAME` with the correct name of the form the WebHook should submit to.

This will allow you to process the form to multiple locations by creating multiple MCRF profiles and WebHooks for a form.

Due to the async processing, I had to revert to the WebHook requestUrl filter to replace the merge tag. Since we need to be able to read the data set in the specific feed, this was the best option. We could use the `gform_{$this->get_slug()}_pre_process_feeds` hook, but you would have to force the aync to false to use this hook. 

I also changed the acronym to match as requested.

rbaugh commented:
Oh, if you have a PHPCS file for standardizing formatting with your setup, please let me know. I was trying to keep it matching, but I am not getting the right setup as you all.

I typically use the WP version like this [phpcs.xml](https://github.com/graydigitalgroup/gf_civicrm/blob/main/phpcs.xml) file.

rbaugh commented:
@agileware-iris When I pushed up a commit, after your last comment, I had already worked on moving the settings to the webhook feed settings, but I believe the other items in your checklist have also been completed.

agileware-iris commented:
@rbaugh Sorry for the delay in getting back to you on this. Thanks for your work on it! We'll merge it in and continue working on the other aspects of the integration from here.
